From 23d8ce4adcadcc5e7474bdd49e32c2a3552006c2 Mon Sep 17 00:00:00 2001
From: ejaaskel <esa.jaaskela@suomi24.fi>
Date: Thu, 30 Oct 2025 12:53:55 +0200
Subject: [PATCH] Add unlocking checking feature

Add device unlocking feature. This can be used to set the device into an
unlocked state, for example, after inputting the correct stop string.
Once in this unlocked state, the device can be allowed to perform
different tasks that are not possible while locked.

Upstream-Status: Pending
---
 common/Kconfig      | 11 +++++++++++
 common/Makefile     |  1 +
 common/unlocking.c  | 14 ++++++++++++++
 include/unlocking.h |  8 ++++++++
 4 files changed, 34 insertions(+)
 create mode 100644 common/unlocking.c
 create mode 100644 include/unlocking.h

diff --git a/common/Kconfig b/common/Kconfig
index 47d17f4e7c6..5f79c0ba916 100644
--- a/common/Kconfig
+++ b/common/Kconfig
@@ -976,6 +976,17 @@ config BOARD_RNG_SEED
 	  Default seed size (64 bytes) can be overridden by a decimal
 	  environment variable rng_seed_size.
 
+config UNLOCKING
+	bool "Enable device unlocking feature"
+	help
+	  This option adds two functions that allow setting the system
+	  into an unlocked state and checking whether the state is
+	  unlocked.
+
+	  This functionality can be used to filter and limit certain
+	  actions if for example the stop string has not been entered
+	  to the system.
+
 endmenu
 
 menu "Update support"
diff --git a/common/Makefile b/common/Makefile
index d62ea34599e..a63e1b72592 100644
--- a/common/Makefile
+++ b/common/Makefile
@@ -15,6 +15,7 @@ obj-$(CONFIG_HUSH_MODERN_PARSER) += cli_hush_modern.o
 obj-$(CONFIG_AUTOBOOT) += autoboot.o
 obj-$(CONFIG_BUTTON_CMD) += button_cmd.o
 obj-y += version.o
+obj-$(CONFIG_UNLOCKING) += unlocking.o
 
 # # boards
 obj-y += board_f.o
diff --git a/common/unlocking.c b/common/unlocking.c
new file mode 100644
index 00000000000..42829d6103f
--- /dev/null
+++ b/common/unlocking.c
@@ -0,0 +1,14 @@
+#include <unlocking.h>
+
+u32 unlock_state = 0;
+
+// Use magic value for simple overwrite protection
+#define UNLOCK_MAGIC 0x41555448
+
+bool device_is_unlocked(void) {
+	return unlock_state == UNLOCK_MAGIC;
+}
+
+void device_set_unlocked(void) {
+	unlock_state = UNLOCK_MAGIC;
+}
diff --git a/include/unlocking.h b/include/unlocking.h
new file mode 100644
index 00000000000..ac3181b4bf6
--- /dev/null
+++ b/include/unlocking.h
@@ -0,0 +1,8 @@
+#ifndef UNLOCKING_H
+#define UNLOCKING_H
+#include <linux/types.h>
+
+bool device_is_unlocked(void);
+
+void device_set_unlocked(void);
+#endif
